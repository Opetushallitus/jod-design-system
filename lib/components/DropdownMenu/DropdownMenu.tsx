import React from 'react';

export interface DropdownMenuOptionsData<T extends string = string> {
  value: T;
  label: string;
}

interface DropdownMenuProps<T extends DropdownMenuOptionsData, U extends string = string> {
  /** Label for the component */
  label: string;
  /** Hide label. Still available for screenreaders */
  hideLabel?: boolean;
  /** Options for component */
  options: T[];
  /** Default value to be selected initially */
  defaultValue?: U;
  /** Controlled mode */
  selected?: U;
  /** Callback on selection change */
  onChange?: (value: U) => void;
  /** Component is disabled for user interaction */
  disabled?: boolean;
}

export const DropdownMenu = <
  U extends string = string,
  T extends DropdownMenuOptionsData<string> = DropdownMenuOptionsData<string>,
>({
  label,
  hideLabel = false,
  options,
  defaultValue,
  onChange: propOnChange,
  disabled = false,
}: DropdownMenuProps<T, U>) => {
  const labelId = React.useId();
  return (
    <div className="ds-flex ds-flex-row">
      {!hideLabel && (
        <label htmlFor={labelId} className="ds-text-jod-black ds-mr-2 ds-self-center ds-font-bold">
          {label}
        </label>
      )}
      <select
        disabled={disabled}
        id={hideLabel ? undefined : labelId}
        aria-label={hideLabel ? label : undefined}
        className="hover:ds-bg-purple-100 focus:ds-ring-purple-500 disabled:ds-border-gray-500 disabled:ds-bg-gray-200 disabled:ds-text-gray-500 disabled:hover:ds-bg-gray-200 ds-border-jod-dark ds-bg-jod-white ds-min-w-[120px] ds-justify-self-end ds-rounded-lg ds-border ds-p-2 focus:ds-outline-none focus:ds-ring"
        onChange={(event: React.ChangeEvent<HTMLSelectElement>) => {
          if (propOnChange) {
            propOnChange(event.target.value as U);
          }
        }}
        defaultValue={defaultValue}
      >
        {options.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
    </div>
  );
};
