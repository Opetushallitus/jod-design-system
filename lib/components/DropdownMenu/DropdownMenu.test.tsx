import { describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { DropdownMenu } from './DropdownMenu';

const options = [
  { value: 'fi', label: 'Suomi' },
  { value: 'sv', label: 'Svenska' },
  { value: 'en', label: 'English' },
];

describe('Snapshot testing', () => {
  it('should render with defaults', () => {
    const { container } = render(<DropdownMenu label="Label" options={options} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not show label', () => {
    const { container } = render(<DropdownMenu hideLabel={true} label="Label" options={options} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render as disabled', () => {
    const { container } = render(<DropdownMenu disabled={true} label="Label" options={options} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultValue', () => {
  it('should select first by default', () => {
    render(<DropdownMenu label="Label" options={options} />);
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    expect((screen.getByRole('option', { name: 'Suomi' }) as HTMLOptionElement).selected).toBe(true);
  });

  it('should select the wanted one', () => {
    render(<DropdownMenu label="Label" options={options} defaultValue="sv" />);
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    expect((screen.getByRole('option', { name: 'Svenska' }) as HTMLOptionElement).selected).toBe(true);
    //
  });

  it('should still allow user to another', async () => {
    render(<DropdownMenu label="Label" options={options} defaultValue="fi" />);
    const user = userEvent.setup();
    await user.selectOptions(screen.getByRole('combobox'), 'English');

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    expect((screen.getByRole('option', { name: 'English' }) as HTMLOptionElement).selected).toBe(true);
  });
});

describe('onChange', () => {
  const options = [
    { value: 'fi', label: 'Suomi' },
    { value: 'sv', label: 'Svenska' },
    { value: 'en', label: 'English' },
  ];

  it('is called on user selection', async () => {
    const mockCallback = vi.fn();
    const user = userEvent.setup();

    render(<DropdownMenu label="Label" options={options} onChange={mockCallback} />);

    await user.selectOptions(screen.getByRole('combobox'), 'English');
    expect(mockCallback).toHaveBeenCalledTimes(1);
  });

  it('is not called when disabled', async () => {
    const mockCallback = vi.fn();
    const user = userEvent.setup();

    render(<DropdownMenu label="Label" options={options} onChange={mockCallback} disabled={true} />);

    await user.selectOptions(screen.getByRole('combobox'), 'English');
    expect(mockCallback).toHaveBeenCalledTimes(0);
  });
});

describe('label', () => {
  it('should render with label', () => {
    render(<DropdownMenu label="Label" options={[]} />);
    expect(screen.getByLabelText('Label')).not.toBeNull();
  });
});

describe('hideLabel', () => {
  it('should add aria-label when label is hidden', () => {
    render(<DropdownMenu label="Non-visible label" options={[]} hideLabel={true} />);
    expect(screen.getByRole('combobox', { name: 'Non-visible label' })).not.toBeNull();
  });
});
