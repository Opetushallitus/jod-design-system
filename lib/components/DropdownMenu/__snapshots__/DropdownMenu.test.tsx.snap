// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > should not show label 1`] = `
<div
  class="ds-flex ds-flex-row"
>
  <select
    aria-label="Label"
    class="hover:ds-bg-purple-100 focus:ds-ring-purple-500 disabled:ds-border-gray-500 disabled:ds-bg-gray-200 disabled:ds-text-gray-500 disabled:hover:ds-bg-gray-200 ds-border-jod-dark ds-bg-jod-white ds-min-w-[120px] ds-justify-self-end ds-rounded-lg ds-border ds-p-2 focus:ds-outline-none focus:ds-ring"
  >
    <option
      value="fi"
    >
      Suomi
    </option>
    <option
      value="sv"
    >
      Svenska
    </option>
    <option
      value="en"
    >
      English
    </option>
  </select>
</div>
`;

exports[`Snapshot testing > should render as disabled 1`] = `
<div
  class="ds-flex ds-flex-row"
>
  <label
    class="ds-text-jod-black ds-mr-2 ds-self-center ds-font-bold"
    for=":r2:"
  >
    Label
  </label>
  <select
    class="hover:ds-bg-purple-100 focus:ds-ring-purple-500 disabled:ds-border-gray-500 disabled:ds-bg-gray-200 disabled:ds-text-gray-500 disabled:hover:ds-bg-gray-200 ds-border-jod-dark ds-bg-jod-white ds-min-w-[120px] ds-justify-self-end ds-rounded-lg ds-border ds-p-2 focus:ds-outline-none focus:ds-ring"
    disabled=""
    id=":r2:"
  >
    <option
      value="fi"
    >
      Suomi
    </option>
    <option
      value="sv"
    >
      Svenska
    </option>
    <option
      value="en"
    >
      English
    </option>
  </select>
</div>
`;

exports[`Snapshot testing > should render with defaults 1`] = `
<div
  class="ds-flex ds-flex-row"
>
  <label
    class="ds-text-jod-black ds-mr-2 ds-self-center ds-font-bold"
    for=":r0:"
  >
    Label
  </label>
  <select
    class="hover:ds-bg-purple-100 focus:ds-ring-purple-500 disabled:ds-border-gray-500 disabled:ds-bg-gray-200 disabled:ds-text-gray-500 disabled:hover:ds-bg-gray-200 ds-border-jod-dark ds-bg-jod-white ds-min-w-[120px] ds-justify-self-end ds-rounded-lg ds-border ds-p-2 focus:ds-outline-none focus:ds-ring"
    id=":r0:"
  >
    <option
      value="fi"
    >
      Suomi
    </option>
    <option
      value="sv"
    >
      Svenska
    </option>
    <option
      value="en"
    >
      English
    </option>
  </select>
</div>
`;
